syntax = "proto3";

option go_package = "github.com/NethermindEth/juno/internal/services/vmrpc";

message GetValueRequest {
  bytes key = 1;
}

message GetValueResponse {
  bytes value = 1;
}

service StorageAdapter {
    rpc GetValue (GetValueRequest) returns (GetValueResponse) {}
}

message VMCallRequest {
    message BlockInfo {
        uint64 number = 1;
        uint64 timestamp = 2;
        uint64 gas_price = 3;
        uint64 sequencer_address = 4;
    }

    bytes root = 1;
    bytes contract_address = 2;
    bytes selector = 3;
    bytes calldata = 4;
    bytes caller_address = 5;
    bytes signature = 6;
    BlockInfo block_info = 7;
}

message VMCallResponse {
    repeated bytes retdata = 1;
}

service VM {
    rpc Call (VMCallRequest) returns (VMCallResponse) {}
}
